app/grandchallenge/algorithms/migrations/0063_algorithminterface_algorithminterfacethroughtable_and_more.py

```diff
@@ -0,0 +1,137 @@
+# Generated by Django 4.2.16 on 2024-11-25 14:40
+import django.db.models.deletion
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import migrations, models
+from django.db.models import Count
+
+
+def create_algorithm_interfaces(apps, _schema_editor):
+    Algorithm = apps.get_model("algorithms", "Algorithm")  # noqa: N806
+    AlgorithmInterface = apps.get_model(  # noqa: N806
+        "algorithms", "AlgorithmInterface"
+    )
+    AlgorithmInterfaceThroughTable = apps.get_model(  # noqa: N806
+        "algorithms", "AlgorithmInterfaceThroughTable"
+    )
+    ComponentInterface = apps.get_model(  # noqa: N806
+        "components", "ComponentInterface"
+    )
+
+    for algorithm in Algorithm.objects.all():
+        inputs = ComponentInterface.objects.filter(
+            pk__in=algorithm.inputs.values_list("pk", flat=True)
+        ).all()
+        outputs = ComponentInterface.objects.filter(
+            pk__in=algorithm.outputs.values_list("pk", flat=True)
+        ).all()
+
+        try:
+            io = AlgorithmInterface.objects.annotate(
+                input_count=Count("inputs", distinct=True),
+                output_count=Count("outputs", distinct=True),
+            ).get(
+                inputs__in=inputs,
+                outputs__in=outputs,
+                input_count=len(inputs),
+                output_count=len(outputs),
+            )
+        except ObjectDoesNotExist:
+            io = AlgorithmInterface.objects.create()
```

Comment:
We cannot use the model manager method that I added here because migrations work with historic model instances, and don't have access to custom manager methods or model methods. Is that ok given that this is a one off migration? Or would it be better if I created a function to do this that I import here and in all other places where I need it (i.e. the forms)? A manager method seemed better than a separate utility function to me, but I'm open for other suggestions.